Document metrics through OpenTelemetry (#65)
kubewarden/docs: src/operator-manual/opentelemetry/01-quickstart.md

# OpenTelemetry

[OpenTelemetry](https://opentelemetry.io/) is a Cloud Native Computing Foundation framework for
observability. It enables your microservices to provide metrics, logs and traces.

> Kubewarden's `policy-server` is instrumented with the OpenTelemetry SDK, reporting data to an

Reviewer: Paving the way for the controller's integration with OpenTelemetry:

```suggestion
Kubewarden's components are instrumented with the OpenTelemetry SDK, reporting data to an
```

OpenTelemetry collector -- called the agent.

> In addition to the OpenTelemetry collector running as an agent in a sidecar, it can also be deployed
> on the cluster as a gateway. In this operational model, the OpenTelemetry collector running as an
> agent reports metrics, logs and traces to another OpenTelemetry collector running as a gateway,
> configured to report this data to third party services, either running inside of the cluster or
> outside.

Reviewer: ```suggestion
The OpenTelemetry collector can be either run as a sidecar agent, or it can be deployed
inside of the cluster as a gateway.

This documentation focuses on deploying the OpenTelemetry agent as a container sidecar. For more information about
the other deployment modes, please refer to OpenTelemetry official documentation.
```

Author: Here I'm trying to put into words that it's possible to:

1. Component -> Agent -> 3rd party services (e.g. datadog). This requires to configure every sidecar with third party credentials.
2. Component -> Gateway -> 3rd party services. This requires to reconfigure the gateway whenever a service needs to change something.
3. Component -> Agent -> Gateway -> 3rd party services: This is what I'm trying to mention here. In this case, the agent is configured to send everything through OTLP to another OpenTelemetry collector (the gateway), that is configured once to report data to third party services.

Reviewer: I see, then I would revisit my suggestion to be something like:

> By following this documentation, we will integrate OpenTelemetry using the following architecture:
>   * Each Pod of the Kubewarden stack will have a OpenTelemetry sidecar
>   * The sidecar receives tracing and monitoring information from the Kubewarden component via the OpenTelemetry Protocol (OTLP)
>   * The OpenTelemetry collector will:
>     - Send the trace events to a central Jaeger instance
>     - Expose Prometheus metrics on port <XXXX>
>
>  For more information about the other deployment modes, please refer to OpenTelemetry official documentation.

    ▪ Using image k8s.gcr.io/ingress-nginx/controller:v1.0.0-beta.3
    ▪ Using image k8s.gcr.io/ingress-nginx/kube-webhook-certgen:v1.0
🔎  Verifying ingress addon...
🌟  The 'ingress' addon is enabled
```

## Install OpenTelemetry

We are going to use the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator)
to manage the automatic injection of the OpenTelemetry Collector sidecar
inside of the PolicyServer pod.

The OpenTelemetry Operator requires [cert-manager](https://cert-manager.io/docs/installation/)
to be installed inside of the cluster.

This can be done with this command:

```console
kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v1.5.3/cert-manager.yaml
kubectl wait --for=condition=Available deployment --timeout=2m -n cert-manager --all
```

Once cert-manager is up and running, the OpenTelemetry operator can be installed in this way:

```console
kubectl apply -f https://github.com/open-telemetry/opentelemetry-operator/releases/latest/download/opentelemetry-operator.yaml
kubectl wait --for=condition=Available deployment --timeout=2m -n opentelemetry-operator-system --all
```

> ## OpenTelemetry integration
> 
> Kubewarden is instrumented to support OpenTelemetry.
> 
> Let's look at the configuration settings for tracing (Jaeger) and exposing and processing metrics
> (Prometheus) in the following sections of the book.

Reviewer: ```suggestion
We can now move to the next chapters of the book to enable application tracing (via Jaeger integration) and application metrics (via Prometheus integration).
```
